src/pudl/glue/epacems_unitid_eia_plant_crosswalk.py

```diff
@@ -0,0 +1,119 @@
+
+
+def extract(ds: Datastore) -> pd.DataFrame:
+    """Extract the EPACEMS-EIA Crosswalk from the Datastore."""
+    with ds.get_zipfile_resource(
+        "epacems_unitid_eia_plant_crosswalk",
+        name="epacems_unitid_eia_plant_crosswalk.zip",
+    ).open("camd-eia-crosswalk-master/epa_eia_crosswalk.csv") as f:
+        return pd.read_csv(f)
+
+
+def transform(
+    epa_eia_crosswalk: pd.DataFrame, generators_entity_eia: pd.DataFrame
+) -> dict[str, pd.DataFrame]:
+    """Clean up the EPACEMS-EIA Crosswalk file and split it into normalized tables.
+
+    Args:
+        epa_eia_crosswalk: The result of running this module's extract() function.
+        generators_entity_eia: The generators_entity_eia table.
+
+    Returns:
+        A dictionary of three normalized DataFrames comprised of the data
+        in the original crosswalk file. EPA plant id to EPA unit id; EPA plant
+        id to EIA plant id; and EIA plant id to EIA generator id to EPA unit
+        id. Includes no nan values.
+    """
+    logger.info("Cleaning up the epacems-eia crosswalk")
+
+    column_rename = {
+        "camd_unit_id": "unit_id_epa",
+        "camd_plant_id": "plant_id_epa",
+        "eia_plant_name": "plant_name_eia",
+        "eia_plant_id": "plant_id_eia",
+        "eia_generator_id": "generator_id",
+    }
+
+    # Basic column rename, selection, and dtype alignment.
+    crosswalk_clean = (
+        epa_eia_crosswalk.pipe(pudl.helpers.simplify_columns)
+        .rename(columns=column_rename)
+        .filter(list(column_rename.values()))
+        .pipe(apply_pudl_dtypes, "eia")
+    )
+
+    # The crosswalk is a static file: there is no year field. The plant_id_eia and
+    # generator_id fields, however, are foreign keys from an annualized table. If the
+    # fast ETL is run (on one year of data) the test will break because the crosswalk
+    # tables with plant_id_eia and generator_id contain values from various years. To
+    # keep the crosswalk in alignment with the available eia data, we'll restrict it
+    # based on the generator entity table which has plant id and generator id.
+    crosswalk_clean = crosswalk_clean.merge(
+        generators_entity_eia[["plant_id_eia", "generator_id"]],
+        on=["plant_id_eia", "generator_id"],
+        how="inner",
+    )
+
+    # There are some eia generator_id values in the crosswalk that don't match the eia
+    # generator_id values in the generators_eia860 table where the foreign keys are
+    # stored. All of them appear to have preceeding zeros. I.e.: 0010 should be 10.
+    # This makes sure to nix preceeding zeros on crosswalk generator ids that are all
+    # numeric. I.e.: 00A10 will stay 00A10 but 0010 will become 10.
+    crosswalk_clean.loc[
+        crosswalk_clean.generator_id.str.contains(r"^0+\d+$"), "generator_id"
+    ] = crosswalk_clean.generator_id.replace({r"^0+": ""}, regex=True)
+
+    # NOTE: still need to see whether unit_id matches up with the values in EPA well!
```

Comment:
Do you mean the values in the EPA CEMS?  Which we can't enforce with a FK constraint since the CEMS data ends up in parquet, not the database.

This *is* something that we *can* check in the nightly builds though, since they'll have all the EPA CEMS and SQLite data available. We should probably organize some tests that are only run as part of the full ETL, like the validation tests are now. Maybe @bendnorman has thoughts on where we should put them or how they should be organized.

Reply:
I was thinking it would be good to see if there were any discrepancies between the CEMS data and the crosswalk. Even though there is no foreign key relationship, we still want users to be able to map the crosswalk onto the relevant dataset and that requires accurate pairing. Right above this comment I did something similar for the eia `generator_id` where I noticed that some values in the crosswalk had extra zeros in front of them. I'd want to do the same quality assurance check with CEMS.

Reply:
This seems like a new check that we should add to the data validation tests that can get run when we've got all the processed outputs available via `tox -e nuke` or in the nightly builds.